UI support for multipart bin cue + Loading popup fix

guytaylor opened the pull request:
This change adds support for bin/cue in folders
- They are treated as a single object now
- Info screen added for bin/cue details
Info screens reworked

guytaylor committed:
first commit, added cuebni helper method. Pre refactor of SDNavigator

guytaylor committed:
Support BinCue
- storing the cue and type in cache
- reworked the walk methods to handle bincues
- Extra info page for BinCue

guytaylor committed:
- Set inital values for bincue sets
- Cleaned up Info Screens
- Cleaned up boot/sd change sequence

guytaylor committed:
fix for when there is no image in the root folder, removed unused paramters for GetFirstFileRecursive

guytaylor committed:
Fix for Loading Popup not staying open long enough. It now closes when the load led flashing stops

morio committed:
Add a timer to close a message box

This will close the message box and open the next screen after a certain
amount of time lapses.

morio commented:
@guytaylor it looks good, but I think the UI message box should close on a timer like this:
https://github.com/ZuluSCSI/ZuluSCSI-firmware/tree/feature/ui-close-message-box
as the blinking global could be set by some other function causing the message box to not close.

morio commented:
@guytaylor I noticed a few things
1. with a bin/cue directory, it states the bt/sec if 512 and gives and LBA number, you could probably just get rid of the page as bytes per sector change depending what is set for the track in the cue file

2. The bin/cue directory scrolls the a bin filename in the Browser under `Item:`, maybe scroll the directory name instead?

3. It would be nice for bin file with the same filename as the cue file to display bin/cue information, and also get rid of the LBA, bt/sec page. This is a different case from a bin/cue directory, it is when a `mycd.bin` and a `mycd.cue` file exist in the same directory. The cue file will only reference one file, but have multiple tracks point to different offsets in the bin file. Also the bin file listed in the cue file is ignored. This last one could be done after this current pull request is merged in.

guytaylor commented:
@morio and thanks for the message box close code, ill merge that into my branch :)

guytaylor committed:
Merge branch 'feature/ui-close-message-box' of https://github.com/ZuluSCSI/ZuluSCSI-firmware into ui-support-for-multipart-bin-cue

guytaylor committed:
Fix ups for Timed Message box merge

guytaylor commented on src/SDNavigator.cpp:
```diff
@@ -3,6 +3,115 @@
+static bool loadCueSheet(const char *cueFile, CUEParser &parser, u_int64_t &cueFileSize)
+{
+    // Use second half of scsiDev.data as the buffer for cue sheet text
+    size_t halfbufsize = sizeof(scsiDev.data) / 2;
```
I did a quick test of replacing all the calls to:

`controlLoop();`

with

```
if (scsiDev.phase == BUS_FREE)
{
     controlLoop();
 }
```

Everythig still seems to work.
Do you think that is enough?

morio commented:
> for 1 and 2 my idea was that that was the info for the currently loaded bin file of the set.
> 
> I noticed after i loaded a bincue that current_image was set to the current bin, so the info on pages 1,3,4 was for the bin and page 2 gets added in with the higher level info about the set.
> 
> So i was just going with that but happy to change it :)
> 
> From your comments, sounds like you’d like it to be more like the browser view, which does show the directory name? (The browser view can also display the cue name instead of the dir but as the cur file can be seen on page two, i think must the dir name makes sense)

@guytaylor 
Yes, I think it makes more sense to show the directory name for the screen "Info (1/4)", and probably get rid of "Info (4/4)" for bin/cue files, but keep them for ISOs and other drive types

guytaylor committed:
Only update UI when bus is free. Display folder for bincue. Don't show geometry page for bincue

PetteriAimonen reviewed (approved):
Looks ok, if you want to add a check for the `strcat` that would be nice to avoid crashing if a long filename is inside a long directoryname.

guytaylor committed:
- Fixed size and total bin issue
- Added checks for overflow in strcats

Also
- On inital load and image load, cueFilename, cueSize and totalBins are now stored on the DeviceMap, so no need to query them in the info screen
- Added a common update method to save these
- Fixed abug where if an initally loaded image wasn't in the imgdir folder, the root folder wasset incorrectly

guytaylor commented:
All issues addressed. 
and some cleanup and a small bug fix I found too

guytaylor commented on src/ZuluSCSI_disk.cpp:
```diff
@@ -1041,9 +1045,14 @@ int scsiDiskGetNextImageName(image_config_t &img, char *buf, size_t buflen)
-                setCurrentFolder(target_idx, path); // patch the path
+                strcpy(buf, path);
+                strcat(buf, "/");
+                strcat(buf, file);
```
The check is there:

```
                strcpy(buf, path);
                if ( (strlen(path) + strlen(file)+  2) >= MAX_FILE_PATH)
                {
                    logmsg("Error. Path too long. Trying to join '", path, "' with '", file, '"');
                    return 0;
                }
                        
                strcat(buf, "/");
                strcat(buf, file);
```
                
Is it not visible in the change? :/
